Pn532 NFC/RFID reader and RFID common elements and Mifare Card

Ellerbach opened the pull request:
Adding PN532 NFC/RFID reader:
* Support for PN532 as a card reader
* Documentation and example
* Full HSU  serial port implementation
* I2C support but not optimal due to current System.Device.I2c implementation
* Limited support for SPI using a GPIO for chipselect and reverse byte arrays to emulate LSB due to current System.Device.Spi implementation

Adding RFID common elements:
* Common data card classes and enums
* Abstract for RFID/NFC communication with any card reader

Adding Mifare Card class:
* Supports fully Mifare Card 1K and fully tested
* Documentation and example

krwq commented on src/devices/Pn532/CommandSet.cs:
```diff
@@ -0,0 +1,50 @@
+        TgGetInitiatorCommand = 0x88,
+        TgResponseToInitiator = 0x90,
+        TgGetTargetStatus = 0x8A,
+
```
nit: extra enter :smile:

krwq commented on src/devices/Pn532/ErrorCode.cs:
```diff
@@ -0,0 +1,42 @@
+        OverCurrentDetected = 0x2D,
+        NADMissing = 0x2E,
+        Unknown = 0xFF,
+
```
ditto

krwq commented on src/devices/Pn532/ListPassive/TargetBaudRate.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <summary>
+        ///  106 kbps type A (ISO/IEC14443 Type A)
+        /// </summary>
+        B106kbpsTypeA = 0x00,
```
Should those values be implied by PollingType?

Ellerbach replied:
Those ones are specific to the ListPassiveTarget function. They are unfortunately not the same as for the AutoPoll one. I would have love to have only 1 type of baud rates.

krwq commented on src/devices/Pn532/OperatingMode.cs:
```diff
@@ -0,0 +1,17 @@
+    /// </summary>
+    public enum OperatingMode
+    {
+        HSU = 0b0000_0000,
```
HSU? Is that SerialPort? (if so, could you rename?)

Ellerbach replied:
It is a bit different. HSU is High-Speed UART. But HSU is usually communicating at higher speed than standard Serial Port.

krwq replied:
Perhaps expand the name to `HighSpeedUart`

krwq commented on src/devices/Pn532/P3.cs:
```diff
@@ -0,0 +1,22 @@
+namespace Iot.Device.Pn532
+{
+    /// <summary>
+    /// The P3 statte
```
nit: state

krwq commented on src/devices/Pn532/P3.cs:
```diff
@@ -0,0 +1,22 @@
+    /// The P3 statte
+    /// </summary>
+    [Flags]
+    public enum P3
```
please expand this name or make it internal since this is far from obvious what this means

Ellerbach replied:
Those are exposed for most of them IO that can be used. For example to trigger another sensor. Will add some more explanations.

krwq replied:
I see, perhaps just rename to `Port3` and similarly for others

krwq commented on src/devices/Pn532/P7.cs:
```diff
@@ -0,0 +1,18 @@
+    /// The P7 state
+    /// </summary>
+    [Flags]
+    public enum P7
```
same

krwq commented on src/devices/Pn532/ParametersFlags.cs:
```diff
@@ -0,0 +1,39 @@
+        // bit 0: fNADUsed Î Use of the NAD information in case of initiator
+        // configuration (DEP and ISO/IEC14443-4
+        // PCD).
+        fNADUsed = 0b0000_0001,
```
can you remove `f` prefix? It doesn't seem to be useful

krwq commented on src/devices/Pn532/ParametersFlags.cs:
```diff
@@ -0,0 +1,39 @@
+    /// Security Access Module parameters
+    /// </summary>
+    [Flags]
+    public enum ParametersFlags
```
should this be called `SecurityAccessModuleParameters` then the comment could be removed

Ellerbach replied:
It's acutally a mix. And the original name from the documentation is not the best choice from my point of view. For example, the RemovePrePostAmble has nothing related to the security. I'll see what I can do to bring clarity and see what I can rename in the main function as well as in this enum.

krwq replied:
I feel it's still better than current name. I'd go with just renaming this according to spec and then extend from there later on (not super high pri though).

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+    public class Pn532 : RfidWriteRead, IDisposable
+    {
+        // Timeout 
+        private const int ReadTimeOut = 500;
```
is this based on spec? if not please make them public properties or move them to something SerialPort specific (haven't read the rest of the code to see how this is used)

Ellerbach replied:
It is actually depending of the communication protocole used. Those 500 ms are higher values than in all modes (except 19.2K and 9.6K HSU where it can be higher). So 'Ill make it public. With this default value.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// <param name="portName">The port name</param>
+        public Pn532(string portName)
+        {
+#if DEBUG
```
I'd rather see this set in a test directly rather than ifdef this.

Ellerbach replied:
I was planning to remove this further. I will add it to the constructor which will allow to set it right at the constructor level. You may not want a debug view if you are in debug mode.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+            // Stop bit : 1 bit,
+            // Baud rate : 115 200 bauds,
+            LogInfo.Log("Opening serial port 115200, Parity.None, 8 bits, StopBits.One", LogLevel.Debug);
+            _serialPort = new SerialPort(portName, 115200, Parity.None, 8, StopBits.One);
```
(optional for the first iteration) it would be useful to create internal abstraction similar to what you've proposed in the issue just for this

Ellerbach replied:
I was thinking of that. But due to the limitation of the current I2C and SPI, it will be useless, they have to be managed in a different way.

krwq replied:
what limitations?

Ellerbach replied:
> what limitations?

Reading and writting are not the same way exactly. Here are the elements:
- for I2c, because we open and close the communication channel every time we read a byte or a byte array, I have to read much more than I should (and it's not possible to know in advance how many bytes will be sent).
- For SPI, there is this issue with the LSB first but that's not what is blocking here, it's just that the protocol is asking for 1 more byte to read, write or check the status.
So for all the 3 way of reading and writting on the sensor, we have to use different logic, even if they are very close to each others.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+
+            WakeUp();
+            // Set the SAM
+            var ret = SetSecurityAccessModule();
```
explicit type please

Ellerbach replied:
and for the 2 constructors as well

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+            //var ret = SetSAM();
+            LogInfo.Log($"Setting SAM changed: {ret}", LogLevel.Info);
+            // Check the version
+            if (!CheckVersion())
```
conventionally `Check*` functions should directly throw and not return anything. Other option is to rename this to `IsPn532Version` or something similar

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        public Pn532(SpiDevice spiDevice)
+        {
+#if DEBUG
+            LogLevel = LogLevel.Debug;
```
ditto

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+            // Apply default parameters
+            ret = SetParameters(ParametersFlags.fAutomaticATR_RES | ParametersFlags.fAutomaticRATS);
+            LogInfo.Log($"Setting Parameters Falgs changed: {ret}", LogLevel.Info);
+            ret = SetSecurityAccessModule();
```
why does this need to be set twice?

Ellerbach replied:
The first one always return a timeout or equivalent. It's the case right after waking it up. So it needs to be reapplied after. Will clean the comment and add this.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        /// <param name="diagnoseMode">The self test to run</param>
+        /// <returns>True when success</returns>
+        public bool RunSelfTest(DiagnoseMode diagnoseMode)
```
I think this should return SelfTestStatus or throw an exception with useful info. I.e. what should user do when this returns false?

Ellerbach replied:
This function will have to evolve once advance tests will be implemented. Rational not to throw exception is to be able to run the tests in a raw and see the results.
Some tests will require to return a more complex object like the PollingTestToTarget one. I will leave it so far as not really the priority for the moment.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        public uint VirtualCardTimeout
+        {
+            get { return _virtualCardTimeout * 50; }
```
should this be wrapped in `checked { ... }`?

Ellerbach replied:
I don't get that point.

krwq replied:
basically add a protection against overflowing _virtualCardTimeout

Ellerbach replied:
> basically add a protection against (accidental) overflowing _virtualCardTimeout

I think it's done by this one:
```csharp
if (value / 50 > 0xFF)
     throw new ArgumentException($"{nameof(VirtualCardTimeout)} can't be more than 12750 milliseconds.");
```

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        public SecurityAccessModuleMode SecurityAccessModuleMode
+        {
+            get { return _securityAccessModuleMode; }
+            set { LogInfo.Log($"{nameof(SecurityAccessModuleMode)} changed: {SetSecurityAccessModule()}", LogLevel.Debug); }
```
this doesn't seem to be doing anything. Could you also split this into multiple lines so that you don't execute anything while formatting?

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        }
+
+
+        private bool SetSecurityAccessModule()
```
i feel like this should take the argument

Ellerbach replied:
It is using the internal variables

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+            return ret >= 0;
+        }
+
+        private bool CheckVersion()
```
as mentioned before: throw or start name with `Is` or something which suggests returning `Yes`/`No` answer

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// <param name="maxTarget">The maximum number of targets</param>
+        /// <param name="targetBaudRate">The baud rate to use</param>
+        /// <returns>A raw byte array with the data of the targets if any has been identified</returns>
+        public byte[] ListPassiveTarget(MaxTarget maxTarget, TargetBaudRate targetBaudRate)
```
I'm not familiar with this API - does targetBaudRate refer to "find me all cards talking with this speed" or rather "use this baud rate to find all cards"? This should be clarified in the comment. Also `byte[]` output is weird - is one card represented by one byte or should this provide some way to enumerate on these bytes (fine to leave this as is for now but please create an issue)

Ellerbach replied:
I was struggling with the byte[] return. The device returns a byte[] with [0] as the target number. The rest are the information related to the card it found. And depending of the type, the length of the array is not the same. Reason why there are the Decode functions. To be able to find out what is returned if you are looking at diverse cards.
This is a bit different than the AutoPoll function which returne [0] the target number [1] the type of card and then the raw data.
So we can make this evolve, feedback welcome once you'll play with it a bit so we can decide the best way to have this handled. And same apply to the AutoPoll function.
For the baud rate, it's the baudrate which will be used to find the card.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        /// <param name="toDecode">The raw byte array</param>
+        /// <returns>A decoded card of null if it cant't</returns>
+        public Data106kbpsTypeA Decode106kbpsTypeA(Span<byte> toDecode)
```
This should be TryDecode...

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+                }
+                return data;
+            }
+            catch (Exception)
```
please only catch exceptions you expect

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        /// <param name="toDecode">The raw byte array</param>
+        /// <returns>A decoded card of null if it cant't</returns>
+        public Data106kbpsTypeB DecodeData106kbpsTypeB(Span<byte> toDecode)
```
same comments as above

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        /// <param name="toDecode">The raw byte array</param>
+        /// <returns>A decoded card of null if it cant't</returns>
+        public Data212_424kbps DecodeData212_424Kbps(Span<byte> toDecode)
```
ditto

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// </summary>
+        /// <param name="toDecode">The raw byte array</param>
+        /// <returns>A decoded card of null if it cant't</returns>
+        public Data106kbpsInnovisionJewel DecodeData106kbpsInnovisionJewel(Span<byte> toDecode)
```
ditto

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1623 @@
+        /// <returns>True is success</returns>
+        public bool SetRfField(RfFieldMode rfFieldMode)
+        {
+            return RfConfiguration(RfConfigurationMode.RfField, new byte[1] { (byte)rfFieldMode });
```
RfConfiguration should have a verb in the name, otherwise it's weird to read this code

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1626 @@
+    public class Pn532 : RfidWriteRead, IDisposable
+    {
+        // Timeout 
+        private const int SerialReadTimeOut = 500;
```
I feel like this class should take SerialPort or Stream (i.e. coming from SerialPort) as an input and not be polluted with serial port stuff

Ellerbach replied:
I'll simplify all this and remove it. From page 34 from the documentation, file:///C:/Repos/PN532/docs/141520.pdf, I can get exact timing based on baud rates. I'll use those for the time out.

Ellerbach replied:
Now, I'll still keep the other properties as it's a different one for SPI and I2C and not very well documented, so din't find a way to make exact ones.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1626 @@
+            WakeUp();
+            // Set the SAM
+            bool ret = SetSecurityAccessModule();
+            //var ret = SetSAM();
```
why is this commented out? is there some implied TODO here?

Ellerbach replied:
Just forgot to remove it

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1626 @@
+            // Check the version
+            if (!IsPn532())
+                throw new Exception("Can't find a PN532");
+            // Apply default parameters
```
nit: enter before comment

Ellerbach replied:
Few ones like that. Will try to fix all of them

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1626 @@
+            _controller.OpenPin(_pin, PinMode.Output);
+            _controller.Write(_pin, PinValue.High);
+            Thread.Sleep(2);
+            LogLevel = LogLevel.Debug;
```
overriding input

Ellerbach replied:
Removed

krwq commented on src/devices/Rfid/Data212_424kbps.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// 212 and 424 card elements
+    /// </summary>
+    public class Data212_424kbps
```
(optional for first iteration) why is kbps in the name? the struct doesn't seem to be related with any speed. Is there any better name than number for this perhaps?

Ellerbach replied:
The data returned are different depending on the speed communication between the card and the module. so kbps is actually defining what the number are. I would keep it for clarity

krwq commented on src/devices/Rfid/LogInfo.cs:
```diff
@@ -0,0 +1,52 @@
+
+        public static LogTo LogTo { get; set; } = LogTo.Console;
+
+        public static void Log(string toLog, LogLevel logLevel)
```
I believe there is a way to get TextWriter or Stream from console - that would be a more robust input than enum

Ellerbach replied:
I found useful to be able to select which output to use. I've been working with containers on RPI with this sensor and using Console was useful. While connected on my Windows machine, Debug was all perfect. I used as well both in one occasion, when debugging SPI on the RPI connected in SSH with VS to do the debug.

krwq commented on src/devices/Rfid/Mifare/Mifare.csproj:
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
```
I feel like we should not subsplit Rfid project. Just one top project seem good enough I think (especially it ends up being in Iot.Device.Bindings anyway)

Ellerbach replied:
I made few modifications as I've been playing as well with a USB smart card reader:
- I renamed the main Rfid folder to Card
- I moved the Rfid elements under a folder called Rfid which is specific to Rfid
- Mifare card can be with a chip as well
- I've been working on a Credit Card project so I'll add the Credit Card for the next commit. This does allow to read all available data from the Credit Card. This part is working as well with a normal USB reader. Just need to pass a reader (rfid or smart card) that implement the CardWriteRead class with the function WriteRead and all is working the same way.
- I regrouped Rfid specific and Card transversal 
Reason why:
- Readers are independ from the cards (Mifare, Credit Card and others we will add later on)
- Readers have a depency on Rfid or Card or both

krwq commented on src/devices/Rfid/Mifare/MifareCard.cs:
```diff
@@ -0,0 +1,516 @@
+                C3 = 0;
+            }
+
+            if (accessType == (AccessType.ReadKeyA | AccessType.ReadKeyB | AccessType.WriteKeyB |
```
should you create a separate flag for this combination? Or should this perhaps be: `accessType == AccessType.ReadKeyA || accessType == ...`

krwq replied:
similar question below

Ellerbach replied:
Unfortunately, only specific conditions adds up. You need the flag to add them but it's not possible to have a simple condition like you describe.

krwq commented on src/devices/Rfid/Mifare/MifareCard.cs:
```diff
@@ -0,0 +1,516 @@
+
+            // Table of truth
+            if ((C1a == 0) && (C2a == 0) && (C3a == 0))
+                return AccessSector.WriteKeyAWithKeyA | AccessSector.ReadAccessBitsWithKeyA | AccessSector.ReadKeyBWithKeyA |
```
is there any way to create those flags directly in such a way that all those ifs are not necessary?

Ellerbach replied:
Unfortunately not. See documentation https://www.nxp.com/docs/en/data-sheet/MF1S70YYX_V1.pdf page 12 and 13

krwq replied:
a bit bummer design on the hardware :smile:

krwq commented on src/devices/Rfid/Mifare/MifareCard.cs:
```diff
@@ -0,0 +1,516 @@
+                    ser[1] = BlockNumber;
+                    return ser;
+            }
+
```
\n

krwq commented:
Looks good, nothing blocking, couple of comments which perhaps will simplify this a bit

shahid-pk commented on src/devices/Pn532/ErrorCode.cs:
```diff
@@ -0,0 +1,41 @@
+        CommandNotAcceptable = 0x27,
+        ConfiguredTargetBeenReleased = 0x29,
+        IDCardDoesNotMAtch = 0x2A,
+        CardDisappeared = 0x2B,
```
nit IDCardDoesNotMAtch -> IDCardDoesNotMatch ?

krwq commented:
@Ellerbach I think you might be missing `<PackageReference Include="System.IO.Ports" Version="4.5.0" />` in Iot.Device.Bindings - I'm unable to build this locally without it. Seems for some reason CI didn't run on this PR cc: @joperezr

joperezr commented:
/azp run

krwq commented:
Also some other feedback:
- projects should be targeting netcoreapp2.1
- one reference in the csproj is invalid (`CreditCardProcessing.csproj` is missing) but it doesn't seem to be needed to build
- you should clean up csprojes in general to match other projects

joperezr commented on src/devices/Rfid/Rfid.csproj:
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
```
NIT: netcoreapp2.1

Ellerbach replied:
Yea, fixed all of them in all projects

Ellerbach commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1626 @@
+            ret = SetParameters(ParametersFlags.AutomaticATR_RES | ParametersFlags.AutomaticRATS);
+            LogInfo.Log($"Setting Parameters Falgs changed: {ret}", LogLevel.Info);
+            ret = SetSecurityAccessModule();
+            //var ret = SetSAM();
```
Removed everywhere.

Ellerbach commented:
I remove the ```Rfid```folder from my Commit as I renamed it to Card and the build is failing because it doesn't find it. I may have miss something somewhere but I don't find where.

krwq commented:
@Ellerbach you got build errors because you are in the inconsistent state: when you created PR we still used package references, now we use project to project references - basically when your PR gets run on the CI you got both and `SystemDeviceGpioPackageVersion` is empty variable.

To fix

- rebase the project against master (`git fetch --all && git rebase dotnet/master` or equivalent)
- replace

```xml
<PackageReference Include="System.Device.Gpio" Version="$(SystemDeviceGpioPackageVersion)" />
```

with

```xml
    <ProjectReference Include="$(MainLibraryPath)System.Device.Gpio.csproj">
      <AdditionalProperties>$(AdditionalProperties);RuntimeIdentifier=linux</AdditionalProperties>
    </ProjectReference>
```

krwq commented on src/devices/Card/CreditCard/ApduCommands.cs:
```diff
@@ -0,0 +1,27 @@
+    /// </summary>
+    public class ApduCommands
+    {
+        static public byte[] ApplicaitonBlock = { 0x80, 0x1E };
```
`Applicaiton` -> `Application`

krwq commented on src/devices/Card/CreditCard/ApduCommands.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// The list of predefined commands to communicate with the card
+    /// </summary>
+    public class ApduCommands
```
these should be either internal or create new instance of array each time or be IEnumerable - otherwise modification of these will have effect on the code

krwq commented on src/devices/Card/CreditCard/ApplicationDataDetail.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// The number of records 
+        /// </summary>
+        public byte NumberRecords { get; set; }
```
NumberOfRecords or RecordCount

krwq commented on src/devices/Card/CreditCard/BerSplitter.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <summary>
+    /// A simple Basic Encoding Rules (defined in ISO/IEC 8825–1) decoder
+    /// </summary>
+    public class BerSplitter
```
internal. cc: @bartonjs

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+    /// </summary>
+    public class CreditCard
+    {
+        private const byte Cla = 0x00;
```
can these be somehow expanded?

Ellerbach replied:
Added a comment

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+        private const byte P2 = 0x03;
+        private const byte Lc = 0x04;
+
+        private const int MaxBuffer = 260;
```
MaxBufferSize

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+        private byte _target;
+
+        // This is a string "2PAY.SYS.DDF01" (PPSE) to select the root directory
+        public readonly byte[] RootDirectory = { 0x32, 0x50, 0x41, 0x59, 0x2e, 0x53, 0x59, 0x53, 0x2e, 0x44, 0x44, 0x46, 0x30, 0x31 };
```
Why not `Encoding.ASCII.GetBytes`? (or whatever encoding that is)

Ellerbach replied:
It's just because it's a readonly, it's used only once to select the root directory, it's part of the specifications and always represented like this. I feel like adding ```Encoding.ASCII.GetBytes``` in the constructor won't really add any value at all.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+        /// </summary>
+        /// <param name="nfc">A compatible Card reader</param>
+        /// <param name="target">The target number as some readers needs it</param>
+        public CreditCard(CardWriteRead nfc, byte target)
```
is this meant to be ever constructed by the users or just used? if not then make it internal if yes then target needs some more so it's clearer what it means and how to get one

Ellerbach replied:
User has the pass the reader he wants to use. Without the reader, there is no real interest using this class. See the example with the PN532. I'm using it as well with a USB SmartCard reader, made a simple wrapper and works perfectly as well.
How would you then describe more and better than in the README? What's missing there then? Being in the project, I may miss some elements to make it more clear.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+        /// Gather all the public information present in the credit card.
+        /// Fill then Tag list with all the found information
+        /// </summary>
+        public void FillCreditCardInformation()
```
summary should have info where is this info available. Also I'd replace `Fill` with `Read`

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+                    {
+
+                        // We need to select the Template 0x6F where the Tag 0x84 contains the same App Id and where we have a template A5 attached. 
+                        var templates = Tags.Where(m => m.TagNumber == 0x6F).Where(m => m.Tags.Where(t => t.TagNumber == 0x84).Where(t => t.Data.SequenceEqual(appIdentifier.Data)) != null).Where(m => m.Tags.Where(t => t.TagNumber == 0xA5) != null);
```
huh? at minimum split or format it in such way that it's clear what's executed in what order

Ellerbach replied:
I know this one is a bit complicated that's the reason why I added the comment. I formated it differently so hopefully it will make it a bit more clear

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+                                if (dol.TagNumber == 0x9F66)
+                                {
+                                    // Select modes to get a maximum of data
+                                    TerminalTransactionQualifier ttq = TerminalTransactionQualifier.MagStripeSupported | TerminalTransactionQualifier.EmvModeSuypported | TerminalTransactionQualifier.EmvContachChipSupported |
```
`EmvContachChipSupported` seems to have a typo

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+                                        TerminalTransactionQualifier.OnlinePinSupported | TerminalTransactionQualifier.SignatureSupported | TerminalTransactionQualifier.ConstactChipOfflinePinSupported |
+                                        TerminalTransactionQualifier.ConsumerDeviceCvmSupported | TerminalTransactionQualifier.IssuerUpdateProcessingSupported;
+                                    // Encode the TTq
+                                    toSend[index] = (byte)((long)ttq >> 16);
```
I think (uint) might be a better choice, also you might want to use `BinaryPrimitives.WriteInt32Little/BigEndian`

Ellerbach replied:
moved to uint, yes, will be enough. I only have 3 bytes to convert, reason why I didn't use the primitives. So I changed the enum as well to uint so I'm adding the last byte which is always 0 and I'm using the BinaryPrimitives converter then.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+                                // Transaction amount
+                                else if (dol.TagNumber == 0x9F02)
+                                {
+                                    // Ask authorisation for the minimum, just to make sure
```
nit: authorization

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,497 @@
+                    break;
+                default:
+                    break;
+            }
```
nit: enter after `}`

krwq commented on src/devices/Card/CreditCard/ProcessError.cs:
```diff
@@ -0,0 +1,81 @@
+            // EMV 4.3 Book 3 page 44
+            if (errorToProcess.Length < 2)
+                ErrorType = ErrorType.Unknown;
+            else if ((errorToProcess[0] == 0x90) && (errorToProcess[0] == 0x90))
```
`(errorToProcess[0] == 0x90) && (errorToProcess[1] == 0x90)`?

Ellerbach replied:
```(errorToProcess[0] == 0x90) && (errorToProcess[1] == 0x00)```
Corrected

krwq commented on src/devices/Card/CreditCard/ProcessError.cs:
```diff
@@ -0,0 +1,81 @@
+            // EMV 4.3 Book 3 page 44
+            if (errorToProcess.Length < 2)
+                ErrorType = ErrorType.Unknown;
+            else if ((errorToProcess[0] == 0x90) && (errorToProcess[0] == 0x90))
```
is it possible to construct short from the first two elements and then convert most of those to switch statement or just cast them directly to enum (except for those few values in the bottom where you check only first byte)

Ellerbach replied:
Yea, that should be douable in some ways. See EMV_v4.3_Book_3page 60 and the following ones. It's jsut a summary but that will give you an idea.

krwq commented on src/devices/Card/CreditCard/ProcessError.cs:
```diff
@@ -0,0 +1,81 @@
+            }
+            else if ((errorToProcess[0] == 0x64) || (errorToProcess[0] == 0x65))
+            {
+                ErrorType = ErrorType.ProcessAbordedExecution;
```
`ProcessAbordedExecution` Aborted?

krwq commented on src/devices/Card/CreditCard/Source.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Iot.Device.Card.CreditCardProcessing
+{
+    /// <summary>
+    /// Th esourvce of a Tag
```
spac eoff by 1

Ellerbach replied:
and source not written correctly :-) fixed.

krwq commented on src/devices/Card/CreditCard/Tag.cs:
```diff
@@ -0,0 +1,116 @@
+        /// <summary>
+        /// Search for a specific tag in a list of Tag including the sub Tags
+        /// </summary>
+        /// <param name="tagToDSearch">The list of tags to search in</param>
```
DSearch?

Ellerbach replied:
typo

krwq commented on src/devices/Card/CreditCard/Tag.cs:
```diff
@@ -0,0 +1,116 @@
+        /// </summary>
+        /// <param name="bcds">The BCD encoded byte array</param>
+        /// <returns>The decoded int</returns>
+        public static uint BcdToInt(byte[] bcds)
```
move this here if existing one doesn't work: https://github.com/dotnet/iot/blob/master/src/devices/Common/Iot/Device/Common/NumberHelper.cs#L17
(also you can use it to implement generic version)

Ellerbach replied:
So I'll add here byte[] convertion and Argument checking. BCD numbers on a byte can't be more than 99 otherwise, you'll get a wrong coded number.

krwq commented on src/devices/Card/CreditCard/TagDetails.cs:
```diff
@@ -0,0 +1,94 @@
+    /// </summary>
+    public class TagDetails : Tag
+    {
+        public TagDetails()
```
what's this used for?

krwq commented on src/devices/Card/CreditCard/TagDetails.cs:
```diff
@@ -0,0 +1,94 @@
+        /// Constructor using an existing Tag
+        /// </summary>
+        /// <param name="tag"></param>
+        public TagDetails(Tag tag)
```
would it make sense to replace this with Clone() and implement IClonable? Is this actually used anywhere?

Ellerbach replied:
I'm using the tag details mainly with the static list containing most of the tag I've been able to find in the specs. I don't necessary need to clone it.

krwq commented on src/devices/Card/CreditCard/TerminalTransactionQualifier.cs:
```diff
@@ -0,0 +1,54 @@
+        // byte 1 byte 2 byte 3 byte 4
+        // byte 1
+        // bit 8: 1 = Mag-stripe mode supported 
+        MagStripeSupported = 0b1000_0000_0000_0000_0000_0000,
```
description mentions 4 bytes but this has bits for 3

Ellerbach replied:
Documentation is like that and the first is the one on the left. the last on the right. The last one on the right is always 0.

Ellerbach replied:
And per the comment a bit before, I've changed for uint and made the last (first depends how you look at it) byte as 0 per the spec

krwq commented on src/devices/Card/CreditCard/WorldCountryCurrency.cs:
```diff
@@ -0,0 +1,284 @@
+        /// List of countries, currency names, currency codes, currency code number and country code number
+        /// Use this list to display a country name when extracting the currency or country code names from the card
+        /// </summary>
+        public static List<CountryCurrency> CountryCurrencies = new List<CountryCurrency>()
```
@tarekgh do you perhaps know if we have this data somewhere already perhaps?

Ellerbach replied:
I searched for it but didn't find. If we have anywhere, happy to reuse.

tarekgh replied:
We have RegionInfo which include some of the information populated here. 

I believe we can get the rest of the information from the underlying OS. 

I don't like hardcoding the information here because this will require us to update the library every time something change in this table. especially currency information is changing all the time. It will be good if we can try populate the information dynamically from other source.


---
In reply to: [308507153](https://github.com/dotnet/iot/pull/598#discussion_r308507153) [](ancestors = 308507153)

Ellerbach replied:
> I don't like hardcoding the information here because this will require us to update the library every time something change in this table. especially currency information is changing all the time. It will be good if we can try populate the information dynamically from other source.

I understand, so I will just remove it. It's mainly to display nicely the information from the card. Used that in a project as RegionInfo does not allow to get a country name from an ISO Code currency code for example. I may then open an issue to be able to get a region based on something else than the culture id or culture string.

Ellerbach replied:
Filled an Issue here: https://github.com/microsoft/referencesource/issues/114#issue-475096082

krwq commented on src/devices/Card/CreditCard/category.txt:
```diff
@@ -0,0 +1 @@
+Card
```
lowercase, also you might want to add `rfid` here since it's related

Ellerbach replied:
made the changes in the other files as well

krwq commented on src/devices/Card/LogInfo.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        public static LogLevel LogLevel { get; set; }
+
+        public static LogTo LogTo { get; set; } = LogTo.Console;
```
I still think this should use TextWriter - you can even make the abstraction similar:

```csharp
static class LogTo
{
  public static readonly TextWriter Console = Console.Out;
  // debug actually doesn't have TextWriter but you can either create one or just not have this option
}
```

Ellerbach replied:
How to log both in the Console and in the Debug then? That's what I'm missing with this solution.

krwq replied:
Some example:

```csharp
using System;
using System.Diagnostics;
using System.IO;
using System.Text;

namespace ConsoleApp196
{
    class MultiTextWriter : TextWriter
    {
        TextWriter[] _writers;
        public MultiTextWriter(params TextWriter[] writers)
        {
            _writers = writers;
        }

        public override Encoding Encoding => Encoding.UTF8;

        public override void Write(char value)
        {
            foreach (var writer in _writers)
            {
                writer.Write(value);
            }
        }
    }

    class DebugWriter : TextWriter
    {
        public override Encoding Encoding => Encoding.UTF8;

        public override void Write(char value)
        {
            Debug.Write(value);
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            TextWriter mw = new MultiTextWriter(Console.Out, new DebugWriter());

            mw.WriteLine("TEST");
        }
    }
}
```

krwq commented on src/devices/Card/Mifare/AccessType.cs:
```diff
@@ -0,0 +1,25 @@
+    [Flags]
+    public enum AccessType
+    {        
+        Never = 0b0000_0000,
```
Never or None? (`Never access type` just doesn't sound right)

Ellerbach replied:
OK, will change for None

Ellerbach replied:
Changed as well in the AccessSector

bartonjs commented on src/devices/Card/CreditCard/Tag.cs:
```diff
@@ -0,0 +1,116 @@
+    /// <summary>
+    /// A Tag class containing part of a card information
+    /// </summary>
+    public class Tag
```
If this is about ASN.1/BER, it should be internal, and/or have a more specific name.

A BER/CER/DER reader/writer should be in the next version of .NET Core, with the current prototype at https://github.com/dotnet/corefxlab/tree/master/src/System.Security.Cryptography.Asn1.Experimental / https://dotnet.myget.org/feed/dotnet-corefxlab/package/nuget/System.Security.Cryptography.Asn1.Experimental

Ellerbach replied:
So I'll use it as soon as it will be release. I had a look at it and seems like it can fully work. @krwq , should I use it even if in preview?

joperezr replied:
@Ellerbach we should probably wait a bit till it is shipped as stable. @bartonjs is the plan to move that to corefx? Will it ship as a OOB package? We currently target nc2.1 with our bindings package and it looks like the current implementation you have does as well, I suppose we intend to keep it that way once we move it to corefx?

bartonjs replied:
> is the plan to move that to corefx?

Actually, it's already in corefx and in use, but as internal.

> Will it ship as a OOB package?

My expectation is that it'll be part of System.Security.Cryptography.Encoding, inbox-only.  But maybe we have enough need that it'll be half-in half-out.

Ellerbach replied:
So I'll evaluate how it can be used and if it fills the need. All objects are not with standard tags, they are all custom ones, so if the decoder is not supporting this, I won't be able to use it..

bartonjs replied:
Aside from ASN.1's Real type, that library should do pretty much everything.  Even with Real it'll understand the tag and length, let you get the payload uninterpreted, and then move to the next thing.

Ellerbach replied:
I've been playing a bit with it. It's missing functions to get just raw data and raw tag numbers. It's very much oriented to extract know types and not much for raw.
That said, all the internal functions are present but not exposed and the way the tag is managed in the code makes it not possible to get the raw tag number.

Simple example below (those are real data that I read from cards):

```csharp
            byte[] span = new byte[] { 0x6F, 0x69, 0x84, 0x0E, 0x32, 0x50, 0x41, 0x59, 0x2E, 0x53, 0x59, 0x53, 0x2E, 0x44, 0x44, 0x46, 0x30, 0x31, 0xA5, 0x57, 0xBF, 0x0C, 0x54, 0x61, 0x28, 0x4F, 0x07, 0xA0, 0x00, 0x00, 0x00, 0x42, 0x10, 0x10, 0x50, 0x0B, 0x43, 0x42, 0x20, 0x43, 0x4F, 0x4D, 0x50, 0x54, 0x41, 0x4E, 0x54, 0x9F, 0x2A, 0x01, 0x03, 0x87, 0x01, 0x01, 0x9F, 0x0A, 0x08, 0x00, 0x01, 0x05, 0x02, 0x00, 0x00, 0x00, 0x00, 0x61, 0x28, 0x4F };
            AsnReader reader = new AsnReader(span, AsnEncodingRules.BER);
            // This gives the tag number 0x0F while
            // Real raw tag number is 0x6F
            var ret = reader.PeekTag();
            // this break content is seen as non valid!
            var data = reader.PeekEncodedValue();
            // This break content is seen as non valid!
            var content = reader.PeekContentBytes();
```

As commented in the code, the raw tag number should be 0x6F, it is Tagged 0x0F. Reason is because when decoding the tag, only 7 bits are kept while the raw value needs 8 bits. See here: https://github.com/dotnet/corefxlab/blob/master/src/System.Security.Cryptography.Asn1.Experimental/System/Security/Cryptography/Asn1/Asn1Tag.cs#L210

So once the tag is decoded, it's not possible to get the raw value of the tag.

And because this value is then interpreted as another tag, in this case "reserved" which is not processed, I get an error when I want to extract the data.

So the functiosn are great for known types but not built for raw usage like in SmartCards for example. That would need to be adapted to support raw tag numbers and raw data.

Ellerbach replied:
I've been playing a bit with it. It's missing functions to get just raw data and raw tag numbers. It's very much oriented to extract know types and not much for raw.
That said, all the internal functions are present but not exposed and the way the tag is managed in the code makes it not possible to get the raw tag number.

Simple example below (those are real data that I read from cards):

```csharp
            byte[] span = new byte[] { 0x6F, 0x69, 0x84, 0x0E, 0x32, 0x50, 0x41, 0x59, 0x2E, 0x53, 0x59, 0x53, 0x2E, 0x44, 0x44, 0x46, 0x30, 0x31, 0xA5, 0x57, 0xBF, 0x0C, 0x54, 0x61, 0x28, 0x4F, 0x07, 0xA0, 0x00, 0x00, 0x00, 0x42, 0x10, 0x10, 0x50, 0x0B, 0x43, 0x42, 0x20, 0x43, 0x4F, 0x4D, 0x50, 0x54, 0x41, 0x4E, 0x54, 0x9F, 0x2A, 0x01, 0x03, 0x87, 0x01, 0x01, 0x9F, 0x0A, 0x08, 0x00, 0x01, 0x05, 0x02, 0x00, 0x00, 0x00, 0x00, 0x61, 0x28, 0x4F };
            AsnReader reader = new AsnReader(span, AsnEncodingRules.BER);
            // This gives the tag number 0x0F while
            // Real raw tag number is 0x6F
            var ret = reader.PeekTag();
            // this break content is seen as non valid!
            var data = reader.PeekEncodedValue();
            // This break content is seen as non valid!
            var content = reader.PeekContentBytes();
```

As commented in the code, the raw tag number should be 0x6F, it is Tagged 0x0F. Reason is because when decoding the tag, only 7 bits are kept while the raw value needs 8 bits. See here: https://github.com/dotnet/corefxlab/blob/master/src/System.Security.Cryptography.Asn1.Experimental/System/Security/Cryptography/Asn1/Asn1Tag.cs#L210

So once the tag is decoded, it's not possible to get the raw value of the tag.

And because this value is then interpreted as another tag, in this case "reserved" which is not processed, I get an error when I want to extract the data.

So the functiosn are great for known types but not built for raw usage like in SmartCards for example. That would need to be adapted to support raw tag numbers and raw data.

bartonjs replied:
The tag VALUE is 0x6F.  The tag NUMBER is 0x0F with the Application class (0x40) and a constructed encoding (0x20).

The "raw value" can be multiple bytes, so it only exposes it via the Encode method.   In fact, the left shift you highlighted was about reading a tag that is composed of multiple bytes (and not involved here)

If PeekEncodedValue / PeekContentBytes are saying that the content is invalid, then it's invalid. In your case here, the array named `span` claims to have a payload length of 0x69 bytes, but the entire array is only 68 (decimal) bytes long.  length exceeds payload => invalid.

Ellerbach replied:
hum… so I am was not awake this morning… Will try all gain. Thanks for the explanations. I had then all wrong. Plus the missing byte in the array.

Ellerbach replied:
SO it finds the value and it's possible to get the content. Now, what is clearly missing is a way to get this raw value directly from the "VALUE" directly as you call it, so in this case 0x6F. In case of longer value, reconstructing it with the "NUMBER" and the application class would be complicated while it can be stored directly as a property.
I realized as well that when you have multiple tag like this one which are not part of the native types, you can't make the cursor move. Only available functions are the Peek ones. So you'll need to use the PeekEncodedValue to get the buffer, get the size, use your original buffer and use another reader to decode what's next. 
I may have miss Something as I missed already few things here. But that looks quite complicated. I guess a function that would make the cursor move and get everything at once will be great.

bartonjs replied:
You can get the encoded tag value back by using Asn1Tag.Encode/TryEncode.  If you need it.

```C#
private static readonly Asn1Tag WhateverThisTagIs = new Asn1Tag(TagClass.Application, 0x1F, isConstructed: true);

...

AsnReader outer = new AsnReader(data, AsnEncodingRules.BER);
AsnReader inner = outer.ReadSequence(WhateverTagThisIs);
// Assuming there's not a value expected to be after this one:
outer.ThrowIfNotEmpty();
...
```

Or, if it's contextual:

```C#
if (reader.PeekTag() == WhateverTagThisIs)
{
    ...
}
else
{
    // Maybe just skip it and move to the next thing?
    // ReadEncodedValue returns the whole next value then moves on.
    // So if the return is ignored, it's Skip.
    reader.ReadEncodedValue();
}

You shouldn't ever have to manually slice the ASN structures, the reader does it all.  Everything in ASN breaks down to a built in thing, just use the overloads that take a tag once you have something other than the Universal class tag for the value.

krwq commented on src/devices/Card/CreditCard/ConversionType.cs:
```diff
@@ -0,0 +1,21 @@
+namespace Iot.Device.Card.CreditCardProcessing
+{
+    /// <summary>
+    /// The type of conversion for the data conversion.
```
what do you mean by `default` here in the description? `default(ConversionType)==BcdToString==0`

Ellerbach replied:
Corrected, default(ConversionType) ==ByteArray

krwq commented on src/devices/Card/CreditCard/ConvertionType.cs:
```diff
@@ -0,0 +1,19 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Iot.Device.Card.CreditCardProcessing
```
this seems like a reduced copy of `ConversionType`

Ellerbach replied:
Absolutely, forgot to remove this redundant file.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+        private byte _target;
+
+        // This is a string "1PAY.SYS.DDF01" (PPSE) to select the root directory
+        private readonly byte[] RootDirectory1 = { 0x31, 0x50, 0x41, 0x59, 0x2e, 0x53, 0x59, 0x53, 0x2e, 0x44, 0x44, 0x46, 0x30, 0x31 };
```
`Encoding.ASCII.GetBytes("foo")`

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+                return new ProcessError(received.Slice(0, TailerSize)).ErrorType;
+            }
+            return ErrorType.Unknown;
+
```
nit: extra enter

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+        /// </summary>
+        /// <param name="nfc">A compatible Card reader</param>
+        /// <param name="target">The target number as some readers needs it</param>
+        /// /// <param name="tailerSize">Size of the tailer, most NFC readers add an extra byte 0x00</param>
```
double `///`

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+        /// Create a Credit Card class
+        /// </summary>
+        /// <param name="nfc">A compatible Card reader</param>
+        /// <param name="target">The target number as some readers needs it</param>
```
please add a remark where do user take this from (unless it's required to read a spec to use this class then fine)

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+        /// <param name="nfc">A compatible Card reader</param>
+        /// <param name="target">The target number as some readers needs it</param>
+        /// /// <param name="tailerSize">Size of the tailer, most NFC readers add an extra byte 0x00</param>
+        public CreditCard(CardWriteRead nfc, byte target, int tailerSize = 3)
```
I think `tailer` might be spelled `trailer` but I'm not a native speaker and translator is not too helpful

Ellerbach replied:
It's coming from Tail, so should be ```tailer```. I'm not native neither, let's see for the feedback of a native speaker.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+                        Tag appLocator = null;
+                        if (ret == ErrorType.ProcessCompletedNormal)
+                        {
+                            // Check if we have an Application File Locator 0x94 in 0x77
```
not sure I follow what the comment means, could you add some spec page here?

Ellerbach replied:
Explanation in EMV 4.3 Book 3, Integrated Circuit Card Specifications for Payment Systems. https://www.emvco.com/emv-technologies/contact/. Not really a specific page, it's about reading a lot to understand how things are related, mainly chapter 5 and chapter 7. I will add in the description then.

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,697 @@
+                            for (byte record = 1; record < 5; record++)
+                            {
+                                // 1 fro 10 is for Application Elementary Files 
+                                for (byte Sfi = 1; Sfi < 11; Sfi++)
```
Sfi should start with lowercase

krwq commented on src/devices/Card/CreditCard/Source.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// The source of a Tag
+    /// </summary>
+    [Flags]
```
I don't see any flags here

Ellerbach replied:
Correct, made the change.

krwq commented on src/devices/Card/ErrorType.cs:
```diff
@@ -0,0 +1,31 @@
+    /// </summary>
+    public enum ErrorType
+    {
+        ProcessCompletedNormal,
```
should the first one be called Success for easier discovery?

Ellerbach replied:
I added ```Success``` then at the end with the same value as ```PorcessCompletedNormal```.

krwq commented on src/devices/Card/Mifare/README.md:
````diff
@@ -0,0 +1,75 @@
+
+```csharp
+byte[] retData = null;
+while ((!Console.KeyAvailable))
````
double brackets

krwq commented on src/devices/Card/ProcessError.cs:
```diff
@@ -0,0 +1,96 @@
+            // EMV 4.3 Book 3 page 60
+            if (errorToProcess.Length < 2)
+                ErrorType = ErrorType.Unknown;
+            else if ((errorToProcess[0] == 0x90) && (errorToProcess[1] == 0x00))
```
this piece of code feels like it could be collapsed to something like:

`(ErrorType)(errorToProcess[0] | (errorToProcess[1] << 8))` + changes to `ErrorType`

Ellerbach replied:
There are paths for errors. And some errors are pure errors, some errors are not pure error, the second byte contains a complementary information. They are grouped by family which can be identified with the first byte. See a quite good and exhaustive list of possible errors, warnings and incorrect data lengths, here: https://www.eftlab.co.uk/knowledge-base/complete-list-of-apdu-responses/

krwq commented on src/devices/Card/README.md:
```diff
@@ -0,0 +1,25 @@
+/// This class has to be implemented in all RFID/NFC readers
+/// So Mifare cards can be used the same way independant of any reader
+/// </summary>
+public abstract class RfidWriteRead
```
I wonder if this should be called something like RfidTransceiver or something along these lines. Current name feels a bit awkward but I'm not super familiar with this to propose better name

Ellerbach replied:
Yes, you're right. Changed everywhere for ```CardTransceiver``` and function ```Transcieve```.

krwq commented on src/devices/Common/Iot/Device/Common/NumberHelper.cs:
```diff
@@ -19,13 +21,31 @@ public static int Bcd2Dec(byte bcd)
         /// <returns>BCD Code</returns>
         public static byte Dec2Bcd(int dec)
         {
+            if (dec > 99)
```
`|| dec < 0` + `ArgumentOutOfRangeException`

krwq commented on src/devices/Pn532/CommandSet.cs:
```diff
@@ -0,0 +1,49 @@
+namespace Iot.Device.Pn532
+{
+    /// <summary>
+    /// All cupported command for the PN532
```
typo in `supported`

krwq replied:
also `commands`

Ellerbach replied:
Yes. I corrected a lot of the grammar and orthography in all the files.

krwq commented on src/devices/Pn532/samples/Program.cs:
```diff
@@ -0,0 +1,240 @@
+﻿
```
copyright header

krwq commented on src/devices/Pn532/RfConfiguration/RfTimeout.cs:
```diff
@@ -0,0 +1,30 @@
+    public enum RfTimeout
+    {
+        None = 0x00,
+        T100µs = 0x01,
```
unicode in the name

krwq commented on src/devices/Pn532/RfConfiguration/MaxRetriesMode.cs:
```diff
@@ -0,0 +1,55 @@
+        /// retry, only one try). The default value of this parameter is 0xFF (infinitely).
+        /// For passive mode, the value is always overruled with 0x02 (two retries). 
+        /// </summary>
+        public byte MxRtyATR { get; set; } = 0x02;
```
since this is public the names here should be expanded (i.e. MaxRetryATR - I don't know what ATR stands for but that ideally should be expanded too)

krwq commented on src/devices/Pn532/RfConfiguration/AnalogSettingsTypeBMode.cs:
```diff
@@ -0,0 +1,34 @@
+    /// type B when configured as PCD.
+    /// When using this command, the host controller has to provide 3 new values
+    /// (ConfigurationData[]) for the following internal registers: 
    /// Except for these two specific settings, the 8 remaining analog settings are the
```
random `\r` here

krwq commented on src/devices/Pn532/RfConfiguration/Analog212_424kbpsMode.cs:
```diff
@@ -0,0 +1,46 @@
+    /// </summary>
+    public class Analog212_424kbpsMode
+    {
+        public byte CIU_RFCfg { get; set; } = 0x69;
```
public class - please expand names

Ellerbach replied:
I would love to. See documentation page 104+ https://www.nxp.com/docs/en/user-guide/141520.pdf. If you have any way to find out what is the long name for it, that would be great. First one is obvious, will adjust but for the others, it's far from it.

krwq replied:
https://www.nxp.com/docs/en/nxp/data-sheets/PN532_C1.pdf page 144 seems to have descriptions. Those are not documented in that spec you pointed to (at least I couldn't find it)

krwq commented on src/devices/Pn532/RfConfiguration/Analog212-424_848kbpsMode.cs:
```diff
@@ -0,0 +1,42 @@
+    /// </summary>
+    public class Analog212_424_848kbpsMode
+    {
+        public byte CIU_RxThreshold212 { get; set; } = 0x85;
```
public class => expand names

Ellerbach replied:
same as previous comment

krwq replied:
https://www.nxp.com/docs/en/nxp/data-sheets/PN532_C1.pdf page 144

krwq commented on src/devices/Pn532/RfConfiguration/Analog106kbpsTypeAMode.cs:
```diff
@@ -0,0 +1,50 @@
+    /// </summary>
+    public class Analog106kbpsTypeAMode
+    {
+        public byte CIU_RFCfg { get; set; } = 0x59;
```
public class => expand names

Ellerbach replied:
same as previous comment

krwq replied:
https://www.nxp.com/docs/en/nxp/data-sheets/PN532_C1.pdf page 144

krwq commented on src/devices/Pn532/README.md:
```diff
@@ -0,0 +1,212 @@
+	Span<byte> read = stackalloc byte[512];
+	int ret = -1;
+	while (ret<0)
+		ret = pn532.ReadDataAsTarget(read);
```
indentation is off

krwq replied:
seems not just here - double check this whole file

Ellerbach replied:
hum, that's weird, all look all ok. It's a regular tab generated by Visual Studio like in any other file. With the all renames, I'll have anyway to replace most of the code from the sample file. So let's see how it will look like.

krwq commented on src/devices/Pn532/Pn532.csproj:
```diff
@@ -0,0 +1,26 @@
+    <ProjectReference Include="$(MainLibraryPath)System.Device.Gpio.csproj">
+      <AdditionalProperties>$(AdditionalProperties);RuntimeIdentifier=linux</AdditionalProperties>
+    </ProjectReference>
+    <PackageReference Include="System.IO.Ports" Version="4.6.0-preview6.19303.8" />
```
this should be using the same version as Iot.Device.Bindings - adjust one or the other

Ellerbach replied:
So "4.5.0" like in the Iot.Device.Bindings. But the 4.5.0 versio is not compatible with Linux serial port. So we should adjust the version in the IoT.Device.Bindigs to be compatible with Linux support. Or did I miss something?
So in short the 4.6.0-preview6.19303.8 or the 4.6.0-preview7.19362.9 or any of the 4.6.0-preview version.

krwq replied:
preview7 is latest for now

krwq commented on src/devices/Card/CreditCard/ConversionType.cs:
```diff
@@ -11,8 +11,8 @@ namespace Iot.Device.Card.CreditCardProcessing
-        BcdToString,
-        ByteArray,
+        ByteArray = 0,
+        BcdToString,        
```
nit: extra spaces in the end

krwq commented:
@Ellerbach [my PR](https://github.com/dotnet/iot/pull/685) could have caused conflicts here. Please rebase your PR and make sure to fix all of the docs issues (previously they might not have been reported and PR might have stayed green)

joperezr commented:
/azp run

Ellerbach commented:
> @Ellerbach [my PR](https://github.com/dotnet/iot/pull/685) could have caused conflicts here. Please rebase your PR and make sure to fix all of the docs issues (previously they might not have been reported and PR might have stayed green)

rebased, fixed all missing comments, pushed a new version

Ellerbach commented:
Still getting an error but not at all related to the code I've commited and pushed. @krwq and @joperezr, please have a look as I don't understand which of my modification impacted the build fail. I may have missed something on the rebase.

joperezr commented:
I see what the issue is, the problem seems to be with some sample projects which is using a wildcard for a package version. I'll fix this.

joperezr commented:
I merged my change so problem should be fixed now.

joperezr commented:
/azp run

krwq commented:
Yes, now seems that all issues or "missing xml doc comment" which is expected

Ellerbach commented:
> Yes, now seems that all issues or "missing xml doc comment" which is expected

So all comments added in all enums, I haven't checked how many but it took me half day to go thru everything. Hopefully all is OK for this PR!

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,708 @@
+        /// </summary>
+        /// <param name="pindigits">The pin in a byte array, between 4 and 8 array length</param>
+        /// <returns>The error status</returns>
+        public ErrorType VerifyPin(ReadOnlySpan<byte> pindigits)
```
Would be useful to have a remark telling if pin 1234 should be:

```csharp
byte[] pin = new byte[] { 1, 2, 3, 4 };
// or
byte[] pin = new byte[] { (byte)'1', (byte)'2', (byte)'3', (byte)'4' };
```

krwq replied:
other alternative is throw exception if wrong one is used

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,708 @@
+            return ErrorType.Unknown;
+        }
+
+        private void FillTagList(List<Tag> theTags, ReadOnlySpan<byte> span, uint parent = 0x00)
```
nit: `theTags` -> `tags`

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,708 @@
+        /// <param name="sfi"></param>
+        /// <param name="numberNecords"></param>
+        /// <returns></returns>
+        public void ReadLogEntries(byte sfi, byte numberNecords)
```
`numberNecords` -> `numberOfRecords`?

krwq commented on src/devices/Card/CreditCard/CreditCard.cs:
```diff
@@ -0,0 +1,708 @@
+            }
+            return ret;
+        }
+
```
nit: extra enter

krwq commented on src/devices/Card/CreditCard/TagList.cs:
```diff
@@ -0,0 +1,307 @@
+        /// </summary>
+        public static List<TagDetails> Tags = new List<TagDetails>()
+        {
+            new TagDetails() { TagNumber =0x06, Source = Source.Icc, Decoder = ConversionType.ByteArray, Description="Object Identifier (OID)", IsTemplate = false, IsDol = false },
```
nit: space after `=` for all here

Ellerbach replied:
285 global replace done. I adjusted as well the one for Description (was missing before and after)

Ellerbach replied:
And found as well one with "'T" which I fixed as well

krwq commented on src/devices/Card/Mifare/MifareCard.cs:
```diff
@@ -0,0 +1,522 @@
+    /// </summary>
+    public class MifareCard
+    {
+
```
nit: extra enter

krwq commented on src/devices/Pn532/P7.cs:
```diff
@@ -0,0 +1,26 @@
+    /// Refer to the detailed documentation page 79 and 80
+    /// </summary>
+    [Flags]
+    public enum Port7
```
seems like all ports have following naming conventions:

PXY
X=port number (X=1 => Port1, X=2 => Port2 ...)
Y=pin number
`PXY=1 << y;`

would it make sense to collapse them into one?

krwq replied:
i.e.: `byte ReadPortA()`

Ellerbach replied:
Yes, there is a pattern but not all sub ports are accessible. Only the ones in the enum are. So to reduce errors and make sure the function will be used correctly, I preferred to use an enum. 
Also keep in mind that reading the values for those 2 ports are done at once in the same function. Implementing something that would read every single port would then call the function multiple times vs only 1 time with the current implementation. And what does cost time is to read or write. So better to group the operations at once.
Also, there will be very very few persons using this feature. It's only if you build your own hardware, where you specifically wants to use the extra GPIO thru this device. I guess when used on an embedded hardware where you can run .NET Core, you won't really need those extra GPIO. Only scenario I saw is when a package device with a relay or equivalent is attached to the PN532 and in this case, you basically ave to control it thru the PN532.
So to keep thing fast and simple, I would stick to the current solution. If we see a need in the future to have individual read or write functions, it will be quite fast to add.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1762 @@
+            LogInfo.Log("Opening serial port 115200, Parity.None, 8 bits, StopBits.One", LogLevel.Debug);
+            _serialPort = new SerialPort(portName, 115200, Parity.None, 8, StopBits.One);
+            // Documentation page 39, serial timeout is 89 ms for 115 200
+            _serialPort.ReadTimeout = 89;
```
How does this compare to ReadTimeout property on this class?

Ellerbach replied:
As explained in the comment of ReadTimeout, it's used only for I2C and SPI. For serial port, the timeout is based on the communication speed with specific tables.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1762 @@
+                    // 0x02 is 424 kbps.
+                    // − Result Length : 1,
+                    // − Result : Number of fails (Maximum 128). 
+                    throw new Exception($"Test {diagnoseMode} not implemented");
```
NotImplementedException

krwq replied:
same below

Ellerbach replied:
> LGTM after fixing the comments, please don't rebase

Made all adjustments, build fully locally, checked all ok and just pushed. So should be all ok with CI/CD build.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1762 @@
+                }
+                return data;
+            }
+            catch (ArgumentOutOfRangeException)
```
if there is an easy fix to make this and below not catch exception to detect errors that would be preferred

Ellerbach replied:
No easy fix, with the function returning the array as it does, there are not that many other options. I would have prefer to have an easy way.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1762 @@
+            _serialPort.Close();
+            switch (baudRate)
+            {
+                case BaudRate.B0009600:
```
nit: you can call
`_serialPort.WriteTimeout = _serialPort.ReadTimeout;`
on the very bottom instead of that

also ReadTimeout seems to be very close to: `Math.Round(10252800.0 / BaudRate)` (equal or 1ms more in case of 230400 where you have to round 44.5)

Ellerbach replied:
Yes for the math. I sticked to what is in the documentation page 39 at least, those are the official values. And added the comment on from where the data is coming from plus the approximation but left the data as they are so it's closer to the values in the documentation.

krwq commented on src/devices/Pn532/Pn532.cs:
```diff
@@ -0,0 +1,1762 @@
+                return ret;
+            }
+            else if (_i2cDevice != null)
+        
```
nit: extra enter

krwq replied:
nit: extra enter

krwq commented on src/devices/Pn532/README.md:
```diff
@@ -0,0 +1,221 @@
+if (retData == null)
+	return;
+
+var decrypted = pn532.TryDecode106kbpsTypeA(retData.AsSpan().Slice(1));
```
please explain why Slice(1) is necessary here or improve APIs so it's not

krwq commented on src/devices/Pn532/README.md:
````diff
@@ -0,0 +1,221 @@
+```csharp
+byte[] retData = null;
+while ((!Console.KeyAvailable))
+{
````
nit: use spaces instead of tabs in this file

krwq commented on src/devices/Pn532/README.md:
```diff
@@ -0,0 +1,221 @@
+
+//Check how many tags and the type
+Console.WriteLine($"Num tags: {retData[0]}, Type: {(PollingType)retData[1]}");
+var decrypted = pn532.TryDecodeData106kbpsTypeB(retData.AsSpan().Slice(3));
```
please explain slicing

krwq commented on src/devices/Pn532/README.md:
```diff
@@ -0,0 +1,221 @@
+
+## PN532 as a target
+
+It's possible to change the PN532 mode to be seen as a target byt another reader. A phone with NFC for example. The bellow example shows how to transform the PN532 into a Credit Card:
```
github is showing last character here in red. I suspect it's not `:` and instead some other unicode.

krwq commented on src/devices/Pn532/RfConfiguration/AnalogSettingsTypeBMode.cs:
```diff
@@ -0,0 +1,51 @@
+    /// type B when configured as PCD.
+    /// When using this command, the host controller has to provide 3 new values
+    /// (ConfigurationData[]) for the following internal registers: 
    /// Except for these two specific settings, the 8 remaining analog settings are the
```
still seeing the \r

Ellerbach replied:
Try to fix it again. Let's see.

krwq commented on src/devices/Pn532/samples/Program.cs:
```diff
@@ -0,0 +1,247 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Iot.Device.Pn532;
```
are all of these used? it seems quite a long list of namespace and since this is sample I'd rather reduce if possible

Ellerbach replied:
Great catch, cleaned.
